src/jit/gentree.h

```diff
@@ -4037,26 +3887,186 @@ struct GenTreeIndir : public GenTreeOp
+    {
+    }
+#endif // DEBUGGABLE_GENTREE
+};
+
 // gtObj  -- 'object' (GT_OBJ). */
 
-struct GenTreeObj : public GenTreeIndir
+struct GenTreeObj : public GenTreeBlk
 {
     CORINFO_CLASS_HANDLE gtClass; // the class of the object
 
-    GenTreeObj(var_types type, GenTreePtr addr, CORINFO_CLASS_HANDLE cls)
-        : GenTreeIndir(GT_OBJ, type, addr, nullptr), gtClass(cls)
+    // If non-null, this array represents the gc-layout of the class.
+    // This may be simply copied when cloning this node, because it is not changed once computed.
+    BYTE* gtGcPtrs;
+
+    // If non-zero, this is the number of slots in the class layout that
+    // contain gc-pointers.
+    __declspec(property(get = GetGcPtrCount)) unsigned gtGcPtrCount;
+    unsigned GetGcPtrCount() const
+    {
+        assert(_gtGcPtrCount != UINT32_MAX);
+        return _gtGcPtrCount;
+    }
+    unsigned _gtGcPtrCount;
+
+    // If non-zero, the number of pointer-sized slots that constitutes the class token.
+    unsigned gtSlots;
+
+    bool IsGCInfoInitialized()
+    {
+        return (_gtGcPtrCount != UINT32_MAX);
+    }
+
+    void SetGCInfo(BYTE* gcPtrs, unsigned gcPtrCount, unsigned slots)
+    {
+        gtGcPtrs      = gcPtrs;
+        _gtGcPtrCount = gcPtrCount;
+        gtSlots       = slots;
+        if (gtGcPtrCount != 0)
+        {
+            // We assume that we cannot have a struct with GC pointers that is not a multiple
+            // of the register size.
+            // The EE currently does not allow this, but it could change.
+            // Let's assert it just to be safe.
+            noway_assert(roundUp(gtBlkSize, REGSIZE_BYTES) == gtBlkSize);
+        }
+        else
+        {
+            // ChangeOper(GT_BLK);
+        }
+    }
+
+    void CopyGCInfo(GenTreeObj* srcObj)
+    {
+        if (srcObj->IsGCInfoInitialized())
+        {
+            gtGcPtrs      = srcObj->gtGcPtrs;
+            _gtGcPtrCount = srcObj->gtGcPtrCount;
+            gtSlots       = srcObj->gtSlots;
+        }
+    }
+
+    GenTreeObj(var_types type, GenTreePtr addr, CORINFO_CLASS_HANDLE cls, unsigned size)
+        : GenTreeBlk(GT_OBJ, type, addr, size), gtClass(cls)
+    {
+        // By default, an OBJ is assumed to be a global reference.
+        gtFlags |= GTF_GLOB_REF;
+        noway_assert(cls != NO_CLASS_HANDLE);
+        _gtGcPtrCount = UINT32_MAX;
+    }
+
+    GenTreeObj(var_types type, GenTreePtr addr, GenTreePtr data, CORINFO_CLASS_HANDLE cls, unsigned size)
+        : GenTreeBlk(GT_STORE_OBJ, type, addr, data, size), gtClass(cls)
     {
         // By default, an OBJ is assumed to be a global reference.
         gtFlags |= GTF_GLOB_REF;
+        noway_assert(cls != NO_CLASS_HANDLE);
+        _gtGcPtrCount = UINT32_MAX;
     }
 
 #if DEBUGGABLE_GENTREE
-    GenTreeObj() : GenTreeIndir()
+    GenTreeObj() : GenTreeBlk()
     {
     }
 #endif
 };
 
+// gtDynBlk  -- 'dynamic block' (GT_DYN_BLK). */
+struct GenTreeDynBlk : public GenTreeBlk
```

Comment:
IMO, It would be useful for these header comments to be more descriptive about the characteristics of each of the types of block objects.